Support minimum and maximum protocol version [1.1.0]

levitte opened the pull request:
If for nothing else, they are needed when doing a regression test

Partially fixes #5661

levitte committed:
Support "-min_protocol" and "-max_protocol" in s_server and s_client

If for nothing else, they are needed when doing a regression test

Partially fixes #5661

levitte committed:
In TLSProxy::Proxy, specify TLSv1.2 as maximum allowable protocol

Partially fixes #5661

levitte commented:
Compared to the result displayed in #5661, we now have a lot less failures (I'm only testing the recipes that fail in #5661):

```
: ; OPENSSL_REGRESSION=$(cd ../_build; pwd) make -j8 test TESTS='test_genrsa test_clienthello test_sslcbcpadding test_sslcertstatus test_sslextension test_sslmessages test_sslrecords test_sslsessiontick test_sslskewith0p test_sslvertol test_tlsextms test_cipherlist test_ssl_new test_ssl_old test_fatalerr test_sslapi'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../../1.1.0 \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines; pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl ../../1.1.0/test/run_tests.pl test_genrsa test_clienthello test_sslcbcpadding test_sslcertstatus test_sslextension test_sslmessages test_sslrecords test_sslsessiontick test_sslskewith0p test_sslvertol test_tlsextms test_cipherlist test_ssl_new test_ssl_old test_fatalerr test_sslapi )
../../1.1.0/test/recipes/15-test_genrsa.t .......... Dubious, test returned 4 (wstat 1024, 0x400)
Failed 4/5 subtests 
../../1.1.0/test/recipes/70-test_clienthello.t ..... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../../1.1.0/test/recipes/70-test_sslcbcpadding.t ... ok
../../1.1.0/test/recipes/70-test_sslcertstatus.t ... ok
../../1.1.0/test/recipes/70-test_sslextension.t .... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/3 subtests 
../../1.1.0/test/recipes/70-test_sslmessages.t ..... ok
../../1.1.0/test/recipes/70-test_sslrecords.t ...... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/11 subtests 
../../1.1.0/test/recipes/70-test_sslsessiontick.t .. ok
../../1.1.0/test/recipes/70-test_sslskewith0p.t .... ok
../../1.1.0/test/recipes/70-test_sslvertol.t ....... ok
../../1.1.0/test/recipes/70-test_tlsextms.t ........ ok
../../1.1.0/test/recipes/80-test_cipherlist.t ...... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../../1.1.0/test/recipes/80-test_ssl_new.t ......... Dubious, test returned 5 (wstat 1280, 0x500)
Failed 5/19 subtests 
../../1.1.0/test/recipes/80-test_ssl_old.t ......... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/7 subtests 
../../1.1.0/test/recipes/90-test_fatalerr.t ........ Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 
../../1.1.0/test/recipes/90-test_sslapi.t .......... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/1 subtests 

Test Summary Report
-------------------
../../1.1.0/test/recipes/15-test_genrsa.t        (Wstat: 1024 Tests: 5 Failed: 4)
  Failed tests:  2-5
  Non-zero exit status: 4
../../1.1.0/test/recipes/70-test_clienthello.t   (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../../1.1.0/test/recipes/70-test_sslextension.t  (Wstat: 512 Tests: 3 Failed: 2)
  Failed tests:  2-3
  Non-zero exit status: 2
../../1.1.0/test/recipes/70-test_sslrecords.t    (Wstat: 512 Tests: 11 Failed: 2)
  Failed tests:  4, 6
  Non-zero exit status: 2
../../1.1.0/test/recipes/80-test_cipherlist.t    (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../../1.1.0/test/recipes/80-test_ssl_new.t       (Wstat: 1280 Tests: 19 Failed: 5)
  Failed tests:  1-2, 6, 8, 17
  Non-zero exit status: 5
../../1.1.0/test/recipes/80-test_ssl_old.t       (Wstat: 256 Tests: 7 Failed: 1)
  Failed test:  3
  Non-zero exit status: 1
../../1.1.0/test/recipes/90-test_fatalerr.t      (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
../../1.1.0/test/recipes/90-test_sslapi.t        (Wstat: 256 Tests: 1 Failed: 1)
  Failed test:  1
  Non-zero exit status: 1
Files=16, Tests=82, 20 wallclock secs ( 0.53 usr  0.04 sys + 15.00 cusr  3.27 csys = 18.84 CPU)
Result: FAIL
make[1]: *** [Makefile:158: _tests] Error 1
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make: *** [Makefile:155: tests] Fel 2
```

mspncp commented:
s/-max-protocol/-max_protocol/g  (PR title and commit message of de6c09a1761a68cee4decac93d3f168690b95ab3)

levitte commented:
> s/-max-protocol/-max_protocol/g  (PR title and commit message of de6c09a1761a68cee4decac93d3f168690b95ab3)

Fixed.  Thanks

levitte commented:
I changed the title because I decided to include all related changes, i.e. a number of test programs are affected in a similar way as well.

levitte committed:
Enhance ssltestlib's create_ssl_ctx_pair to take min and max proto version

Have all test programs using that function specify those versions.
Additionally, have the remaining test programs that use SSL_CTX_new
directly specify at least the maximum protocol version.

levitte commented:
With that additional commit, the regression failures are reduced further:

```
: ; OPENSSL_REGRESSION=$(cd ../_build; pwd) make -j8 test TESTS='test_genrsa test_clienthello test_sslcbcpadding test_sslcertstatus test_sslextension test_sslmessages test_sslrecords test_sslsessiontick test_sslskewith0p test_sslvertol test_tlsextms test_cipherlist test_ssl_new test_ssl_old test_fatalerr test_sslapi'
make depend && make _tests
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make[1]: Entering directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
( cd test; \
  mkdir -p test-runs; \
  SRCTOP=../../1.1.0 \
  BLDTOP=../. \
  RESULT_D=test-runs \
  PERL="/usr/bin/perl" \
  EXE_EXT= \
  OPENSSL_ENGINES=`cd .././engines; pwd` \
  OPENSSL_DEBUG_MEMORY=on \
    /usr/bin/perl ../../1.1.0/test/run_tests.pl test_genrsa test_clienthello test_sslcbcpadding test_sslcertstatus test_sslextension test_sslmessages test_sslrecords test_sslsessiontick test_sslskewith0p test_sslvertol test_tlsextms test_cipherlist test_ssl_new test_ssl_old test_fatalerr test_sslapi )
../../1.1.0/test/recipes/15-test_genrsa.t .......... ok
../../1.1.0/test/recipes/70-test_clienthello.t ..... ok
../../1.1.0/test/recipes/70-test_sslcbcpadding.t ... ok
../../1.1.0/test/recipes/70-test_sslcertstatus.t ... ok
../../1.1.0/test/recipes/70-test_sslextension.t .... ok
../../1.1.0/test/recipes/70-test_sslmessages.t ..... ok
../../1.1.0/test/recipes/70-test_sslrecords.t ...... Dubious, test returned 2 (wstat 512, 0x200)
Failed 2/11 subtests 
../../1.1.0/test/recipes/70-test_sslsessiontick.t .. ok
../../1.1.0/test/recipes/70-test_sslskewith0p.t .... ok
../../1.1.0/test/recipes/70-test_sslvertol.t ....... ok
../../1.1.0/test/recipes/70-test_tlsextms.t ........ ok
../../1.1.0/test/recipes/80-test_cipherlist.t ...... skipped: This test recipe isn't support when test regression
../../1.1.0/test/recipes/80-test_ssl_new.t ......... Dubious, test returned 1 (wstat 256, 0x100)
Failed 1/19 subtests 
../../1.1.0/test/recipes/80-test_ssl_old.t ......... ok
../../1.1.0/test/recipes/90-test_fatalerr.t ........ ok
../../1.1.0/test/recipes/90-test_sslapi.t .......... ok

Test Summary Report
-------------------
../../1.1.0/test/recipes/70-test_sslrecords.t    (Wstat: 512 Tests: 11 Failed: 2)
  Failed tests:  4, 6
  Non-zero exit status: 2
../../1.1.0/test/recipes/80-test_ssl_new.t       (Wstat: 256 Tests: 19 Failed: 1)
  Failed test:  2
  Non-zero exit status: 1
Files=16, Tests=81, 21 wallclock secs ( 0.56 usr  0.04 sys + 14.96 cusr  3.59 csys = 19.15 CPU)
Result: FAIL
make[1]: *** [Makefile:158: _tests] Error 1
make[1]: Leaving directory '/home/levitte/gitwrk/openssl.net/official/_build-1.1.0'
make: *** [Makefile:155: tests] Fel 2
```

levitte committed:
fixup! Enhance ssltestlib's create_ssl_ctx_pair to take min and max proto version

mattcaswell commented on apps/apps.h:
```diff
@@ -250,6 +250,8 @@ int set_cert_times(X509 *x, const char *startdate, const char *enddate,
             "Elliptic curve used for ECDHE (server-side only)" }, \
         {"cipher", OPT_S_CIPHER, 's', "Specify cipher list to be used"}, \
+        {"min_protocol", OPT_S_MINPROTO, 's', "Specify the minimum protocol version to be used"}, \
+        {"max_protocol", OPT_S_MAXPROTO, 's', "Specify the maximum protocol version to be used"}, \
```
These options are not documented

mattcaswell replied:
Oh. I take it back. They kind of are:

`In addition to the options below the B<s_client> utility also supports the
common and client only options documented in the
in the "Supported Command Line Commands" section of the L<SSL_CONF_cmd(3)>
manual page.`

levitte replied:
Note that there are still options found in `SSL_CONF_cmd(3)` that are not yet supported in `s_client` or `s_server`.  We should look into those...

levitte commented:
Merged.

1f71819af527883d2438cfc7f6cd9962eb155155 Support "-min_protocol" and "-max_protocol" in s_server and s_client
b4ea929d51c6d0ec0aab78072b81f39289b6de84 In TLSProxy::Proxy, specify TLSv1.2 as maximum allowable protocol
b402b77da33cc36ee893fa498be2e4220178524b Enhance ssltestlib's create_ssl_ctx_pair to take min and max proto version
